tests/conformance/test_conformance.py

```diff
@@ -0,0 +1,449 @@
+    "storage.buckets.getIamPolicy": [get_iam_policy],
+    "storage.buckets.insert": [create_bucket],
+    "storage.buckets.list": [list_buckets],
+    "storage.buckets.lockRententionPolicy": [],  # lock_retention_policy
+    "storage.buckets.testIamPermission": [get_iam_permissions],
+    "storage.notifications.delete": [delete_notification],
+    "storage.notifications.get": [get_notification],
+    "storage.notifications.list": [list_notifications],
+    "storage.objects.get": [get_blob],
+    "storage.objects.list": [list_blobs],  # S1 end
+    "storage.buckets.patch": [patch_bucket],  # S2 start
+    "storage.buckets.setIamPolicy": [],
+    "storage.buckets.update": [update_bucket],
+    "storage.objects.compose": [],  # compose_blob
+    "storage.objects.copy": [copy_blob, rename_blob],
+    "storage.objects.delete": [delete_blob],
+    "storage.objects.insert": [upload_from_string],
+    "storage.objects.patch": [patch_blob],
+    "storage.objects.rewrite": [rewrite_blob],
+    "storage.objects.update": [update_blob],  # S2 end
+    "storage.notifications.insert": [create_notification],  # S4
+}
+
+########################################################################################################################################
+### Helper Methods for Populating Resources ############################################################################################
+########################################################################################################################################
+
+
+def _populate_resource_bucket(client, resources):
+    bucket = client.bucket(uuid.uuid4().hex)
+    client.create_bucket(bucket)
+    resources["bucket"] = bucket
+
+
+def _populate_resource_object(client, resources):
+    bucket_name = resources["bucket"].name
+    bucket = client.get_bucket(bucket_name)
+    blob = bucket.blob(uuid.uuid4().hex)
+    blob.upload_from_string("hello world")
+    blob.reload()
+    resources["object"] = blob
+
+
+def _populate_resource_notification(client, resources):
+    bucket_name = resources["bucket"].name
+    bucket = client.get_bucket(bucket_name)
+    notification = bucket.notification()
+    notification.create()
+    notification.reload()
+    resources["notification"] = notification
+
+
+def _populate_resource_hmackey(client, resources):
+    hmac_key, secret = client.create_hmac_key(
+        service_account_email=_CONF_TEST_SERVICE_ACCOUNT_EMAIL,
+        project_id=_CONF_TEST_PROJECT_ID,
+    )
+    resources["hmac_key"] = hmac_key
+
+
+resource_mapping = {
+    "BUCKET": _populate_resource_bucket,
+    "OBJECT": _populate_resource_object,
+    "NOTIFICATION": _populate_resource_notification,
+    "HMAC_KEY": _populate_resource_hmackey,
+}
+
+
+def _populate_resources(client, json_resource):
+    resources = {}
+
+    for r in json_resource:
+        func = resource_mapping[r]
+        func(client, resources)
+
+    return resources
```

Comment:
Is there any reason why we shouldn't just create the bucket, blob, notification, and HMAC key at the beginning of the test run, rather than for each testcase?

Reply:
Individual test cases may modify/delete the resource. Since it's quick and cheap to create these resources in the emulator, there's not really a good reason to try to deduplicate in an ad-hoc way.